src/OgAccess.php

```diff
@@ -253,6 +281,86 @@ public function userAccessEntity($operation, EntityInterface $entity, AccountInt
+   * @return \Drupal\Core\Access\AccessResult
+   *   The access result object.
+   *
+   * @see \Drupal\og\PermissionManager::getEntityOperationPermissions()
+   */
+  public function userAccessGroupContentEntityOperation($operation, EntityInterface $group_entity, EntityInterface $group_content_entity, AccountInterface $user = NULL) {
+    // Default to the current user.
+    $user = $user ?: $this->accountProxy->getAccount();
+
+    // Check if the user owns the entity which is being operated on.
+    $is_owner = $group_content_entity instanceof EntityOwnerInterface && $group_content_entity->getOwnerId() == $user->id();
+
+    // Retrieve the group content entity operation permissions.
+    $group_entity_type_id = $group_entity->getEntityTypeId();
+    $group_bundle_id = $group_entity->bundle();
+    $group_content_bundle_ids = [$group_content_entity->getEntityTypeId() => [$group_content_entity->bundle()]];
+
+    $permissions = $this->permissionManager->getDefaultEntityOperationPermissions($group_entity_type_id, $group_bundle_id, $group_content_bundle_ids);
+
+    // Filter the permissions by operation and ownership.
+    // If the user does not own the group content, only the non-owner permission
+    // is relevant (for example 'edit any article node'). However when the user
+    // _is_ the owner, then both permissions are relevant: an owner will have
+    // access if they either have the 'edit any article node' or the 'edit own
+    // article node' permission.
+    $ownerships = $is_owner ? [FALSE, TRUE] : [FALSE];
+    $permissions = array_filter($permissions, function (GroupContentOperationPermission $permission) use ($operation, $ownerships) {
+      return $permission->getOperation() === $operation && in_array($permission->getOwner(), $ownerships);
+    });
+
+    // @todo This doesn't really vary by user but by the user's role inside of
+    //   the group. We should create a cache context for OgRole entities.
+    // @see https://github.com/amitaibu/og/issues/219
+    $cacheable_metadata = new CacheableMetadata();
+    $cacheable_metadata->addCacheableDependency($group_content_entity);
+    if ($user->id() == $this->accountProxy->id()) {
+      $cacheable_metadata->addCacheContexts(['user']);
+    }
+
+    if ($permissions) {
+      // Request the user's membership. If the user is not a member, construct a
+      // 'non-member membership' in which the user has the 'anonymous' role. We
+      // can then use this to retrieve the permissions that the user has within
+      // the group, regardless if they are a member or not.
+      $states = [
+        OgMembershipInterface::STATE_ACTIVE,
+        OgMembershipInterface::STATE_PENDING,
+        OgMembershipInterface::STATE_BLOCKED,
+      ];
+      $membership = Og::getMembership($group_entity, $user, $states) ?: $this->getNonMemberMembership($group_entity, $user);
+      foreach ($permissions as $permission) {
+        if ($membership->hasPermission($permission->getName())) {
```

Comment:
I did this following your recommendation in https://github.com/amitaibu/og/pull/217#discussion_r73567630 - I had split off the creation of the membership to a separate method for readability. I made that method protected to prevent people from calling it and saving the membership that it returns.

_Anyway_ you are completely correct, we totally need to replace it with a call to `::userAccess()` here, because otherwise the alter hooks would not fire! I somehow missed that little detail :confounded: